### PR TITLE
DAOS-15287 test: Add register tearDown steps for job managers

### PR DESCRIPTION
**Pull request by phender**
Register methods to stop potentially running mpirun/orterun commands issued by tests during tearDown.

Skip-unit-tests: true
Skip-fault-injection-test: true
Allow-unstable-test: true
Test-tag: pr Hdf5 LlnlMpi4py Romio AggregationPunching EcodMdtest RebuildMdtest RbldWidelyStriped CheckCsumMetricsMdtest EcodOnlineRebuildMdtest osa BasicCheckout MdtestEasy MdtestHard datamover IorCrash DaosPerf DaosPerfLarge DaosRacerParallelTest MacsioTest IorCrash NvmeIoVerification

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/util/job_manager_utils.py [hunk 1]
```diff
@@ -60,19 +63,56 @@ def get_job_manager(test, class_name=None, job=None, subprocess=None, mpi_type=N
...
+    Use this method to ensure the cleanup of processes started by JobManager.run() are handled in
+    the correct order in the test tearDown() method.
+
+    Args:
+        test (Test): the test to which the job manager will be added
+        class_name (_type_): _description_
+        job (_type_): _description_
+        subprocess (_type_): _description_
+        mpi_type (_type_): _description_
+        timeout (_type_): _description_
```
Comment (daltonbohning) [inline src/tests/ftest/util/job_manager_utils.py:83]: docstring

File: src/tests/ftest/util/job_manager_utils.py [hunk 2]
```diff
@@ -60,19 +63,56 @@ def get_job_manager(test, class_name=None, job=None, subprocess=None, mpi_type=N
...
-        if isinstance(test.job_manager, list):
-            test.job_manager.append(job_manager)
-        else:
-            test.job_manager = job_manager
+        job_manager = add_job_manager(test, class_name, job, subprocess, mpi_type, timeout)
 
     return job_manager
 
 
+def add_job_manager(test, class_name, job, subprocess, mpi_type, timeout):
```
Comment (daltonbohning) [inline src/tests/ftest/util/job_manager_utils.py:71]: Why have two methods for getting a job manager?

File: src/tests/ftest/util/data_mover_test_base.py [hunk 3]
```diff
@@ -288,7 +247,8 @@ def new_posix_test_path(self, shared=False, create=True, parent=None, mount_dir_
             cmd = f"sudo mount -t tmpfs none '{path}' -o size={mount_dir_size}"
             if not run_remote(self.log, self.hostlist_clients, cmd).passed:
                 self.fail(f"Failed to mount directory {path}")
-            self.mounted_posix_test_paths.append(path)
+            self.register_cleanup(
+                cleanup_mounted_path, test=self, hosts=self.hostlist_clients, path=path)
```
Comment (daltonbohning) [inline src/tests/ftest/util/data_mover_test_base.py:251]: We need to do this before removing the directory

**Review by daltonbohning — COMMENTED** [review]
We should probably include `ior` and `mdtest` in test tags, especially considering we're trying to branch ASAP

File: src/tests/ftest/util/data_mover_test_base.py (hunk 3 above)
Comment (daltonbohning) [reply src/tests/ftest/util/data_mover_test_base.py:251]: Sorry, I read this backward

Comment (daltonbohning) [reply src/tests/ftest/util/data_mover_test_base.py:251]: I was looking into why this failed
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14245/7/artifact/Functional%20on%20EL%208.8/datamover/negative_space.py/job.log
```
2024-05-16 00:18:14,546 test             L0411 DEBUG| [7] Register: Calling cleanup method cleanup_mounted_path(test=1-./datamover/negative_space.py:DmvrNegativeSpaceTest.test_dm_negative_space_dcp;run-container-dcp-client_processes-hosts-ior-pool_large-pool_small-server_config-engines-0-storage-0-981d, hosts=wolf-105vm3, path=/mnt/share/tmp.2hS8f75viA/test_dm_negative_space_dcp0)
2024-05-16 00:18:14,546 run_utils        L0410 DEBUG| Running on wolf-105vm3 with a 120 second timeout: rm -rf '/mnt/share/tmp.2hS8f75viA/test_dm_negative_space_dcp0'
2024-05-16 00:18:14,925 run_utils        L0264 DEBUG|   wolf-105vm3 (rc=1): rm: cannot remove '/mnt/share/tmp.2hS8f75viA/test_dm_negative_space_dcp0': Device or resource busy
2024-05-16 00:18:14,926 run_utils        L0410 DEBUG| Running on wolf-105vm3 with a 120 second timeout: sudo umount -f '/mnt/share/tmp.2hS8f75viA/test_dm_negative_space_dcp0'

```

File: src/tests/ftest/util/data_mover_test_base.py [hunk 4]
```diff
@@ -22,6 +22,43 @@
...
+        hosts (NodeSet): hosts from which to unmount the paths
+        path (str): the mounted posix test path to cleanup
+
+    Returns:
+        list: a list of any errors detected when removing the pool
+    """
+    # need to remove contents before umount
+    error_list = cleanup_path(test, hosts, path)
+    result = run_remote(test.log, hosts, f"sudo umount -f '{path}'")
+    if not result.passed:
```
Comment (daltonbohning) [inline src/tests/ftest/util/data_mover_test_base.py:39]: Need to unmount and then remove

File: src/tests/ftest/util/job_manager_utils.py (hunk 2 above)
Comment (phender) [reply src/tests/ftest/util/job_manager_utils.py]: The original thought was to move to a more direct method than the existing method, but I have changed my mind.

File: src/tests/ftest/util/data_mover_test_base.py (hunk 4 above)
Comment (phender) [reply src/tests/ftest/util/data_mover_test_base.py]: We need a `cleanup_path(test, hosts, f'path/*'`) call, umount, and then a `cleanup_path(test, hosts, path)` call.

File: src/tests/ftest/util/data_mover_test_base.py [hunk 5]
```diff
@@ -282,13 +237,22 @@ def new_posix_test_path(self, shared=False, create=True, parent=None, mount_dir_
             if not run_remote(self.log, self.hostlist_clients, cmd).passed:
                 self.fail(f"Failed to mkdir {path}")
 
+        # Remove the path during test tearDown. Not needed when using the path as a mount point as
+        # cleanup_path() is called as part of cleanup_mounted_path()
```
Comment (daltonbohning) [inline src/tests/ftest/util/data_mover_test_base.py:241]: nit - I don't think we need a "meta" comment here. I would expect `cleanup_mounted_path` to be self-contained. I.e. I wouldn't think to call BOTH `cleanup_path` and `cleanup_mounted_path` on the same path

**Review by daltonbohning — APPROVED** [review]
LGTM

**Comment by phender** [comment]
All tests passed in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14245/9/testReport/ except:
- daos_perf/small.py:DaosPerf.test_small
- daos_perf/large.py:DaosPerfLarge.test_large
 
Which failed for known issue https://daosio.atlassian.net/browse/DAOS-15687

**Review by dinghwah — APPROVED** [review]

**Pull request merged**